compiler/IREmitter/Payload/patches/vm_insnhelper.c

```diff
@@ -481,3 +484,232 @@ void sorbet_throwReturn(rb_execution_context_t *ec, VALUE retval) {
+    volatile VALUE bodyException = Qnil;
+    volatile VALUE handlerException = Qnil;
+    volatile enum {
+        RunningBody,
+        RunningHandlers,
+    } state = RunningBody;
+    volatile enum ruby_tag_type nleType;
+
+    EC_PUSH_TAG(ec);
+
+    if ((nleType = EC_EXEC_TAG()) == TAG_NONE) {
+    execute_body:
+        state = RunningBody;
+        // tag.state will have been reset appropriately if we got here via `retry`,
+        // but not this variable.
+        nleType = TAG_NONE;
+
+        // Clear out the local variable shared across exception handling regions
+        // where we store the current exception value.
+        sorbet_writeLocal(cfp, exceptionValueIndex, exceptionValueLevel, Qnil);
+
+        // We're also done with whatever exceptions might have gotten thrown along the way.
+        bodyException = Qnil;
+        handlerException = Qnil;
+
+        // Likewise with any results we accumulated along the way.
+        executionResult = Qundef;
+
+        // Run the body.
+        executionResult = body(pc, methodClosure, cfp);
+
+        // If we get to this point, then we know the body has succeeded without throwing
+        // an exception.  We may need to run the `else` handler if the body didn't
+        // return a value that we need to propagate.
+        if (executionResult == Qundef) {
+            state = RunningHandlers;
+
+            executionResult = elseClause(pc, methodClosure, cfp);
+        }
+        rb_set_errinfo(previousException);
+    } else {
+        // If we get here, setjmp has returned a non-zero value.
+
+        // This case is the "obvious" case: something in the body threw; we need to handle
+        // exceptions directly here.
+        if (state == RunningBody) {
+            // We're handling things very similarly to rb_rescue2/rb_vrescue2:
+            // <add-link-here>
+            //
+            // The significant difference from that function is that we're handling all
+            // the non-local exits directly.
+
+            // Whatever kind of non-local exit we have, we need to make sure that
+            // the Ruby control frame for the handler (or the ensure) we're going
+            // to run lives directly under whatever frame we started this process with.
+            rb_vm_rewind_cfp(ec, cfp);
+
+            if (nleType == TAG_RAISE) {
+                // rb_rescue2/rb_vrescue2 would check ec->errinfo here to determine if it
+                // was the "right" kind of error.  Sorbet has already generated code to check
+                // the exception value for us, we can dispatch directly to the handlers here,
+                // which avoids a little bit of overhead.  But we do need to tell the
+                // handlers what the exception value *is*.
+                bodyException = ec->errinfo;
+            } else {
+                // Any other kind of non-local exit will skip the rescue/else handlers.
+                goto execute_ensure;
+            }
```

Comment:
```suggestion
            if (nleType != TAG_RAISE) {
                // Any other kind of non-local exit will skip the rescue/else handlers.
                goto execute_ensure;
            }
            
            // rb_rescue2/rb_vrescue2 would check ec->errinfo here to determine if it
            // was the "right" kind of error.  Sorbet has already generated code to check
            // the exception value for us, we can dispatch directly to the handlers here,
            // which avoids a little bit of overhead.  But we do need to tell the
            // handlers what the exception value *is*.
            bodyException = ec->errinfo;
```
Slight preference for guarding the goto rather than hiding it in the `else`.